### PR TITLE
Generate a config file from env vars and template

### PR DESCRIPTION
**Pull request by mans0954**
Modify the Dockerfile so that a config file is generated from environment variables and a template on start up.

Supports: https://github.com/overleaf/issues/issues/3680

The template is a PoC - I imagine we'll need to add more environment variables for a working solution.

CC: @gh2k

### CODE REVIEW

**Comment by JuneKelly** [comment]
One thing to note, the [GNU envsubst](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html), seems to be a different program from https://github.com/a8m/envsubst, the latter of which is preferable, as it supports default values for variables that are not defined.

**Comment by JuneKelly** [comment]
@mans0954 I've added a commit, which:

- expands the coverage to all variables
- vendors the `envsubst` binary from https://github.com/a8m/envsubst, as it supports default values, which the gnu binary does not
  - (happy to discuss an alternative to this)
- cleans up the dockerfile a bit

**Comment by JuneKelly** [comment]
@gh2k any thoughts on the above?  ^

**Comment by JuneKelly** [comment]
Sibling PR for the dev-env: https://github.com/overleaf/dev-environment/pull/472

**Comment by gh2k** [comment]
> @gh2k any thoughts on the above? ^

Looks like a good solution to me.

File: conf/envsubst_template.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "port": ${GIT_BRIDGE_PORT:-8000},
+    "rootGitDirectory": "${GIT_BRIDGE_ROOT_DIR:-/tmp/wlgb}",
+    "apiBaseUrl": "${GIT_BRIDGE_API_BASE_URL:-https://localhost/api/v0}",
+    "username": "${GIT_BRIDGE_USERNAME}",
+    "password": "${GIT_BRIDGE_PASSWORD}",
+    "postbackBaseUrl": "${GIT_BRIDGE_POSTBACK_BASE_URL:-https://localhost}",
+    "serviceName": "${GIT_BRIDGE_SERVICE_NAME:-Overleaf}",
+    "oauth2": {
+      "oauth2ClientID": "${GITBRIDGE_OAUTH2_CLIENT_ID}",
```
Comment (mans0954) [inline conf/envsubst_template.json]: I think we should use either `GITBRIDGE_` or `GIT_BRIDGE_` throughout.

**Review by mans0954 — COMMENTED** [review]
> * vendors the `envsubst` binary from https://github.com/a8m/envsubst, as it supports default values, which the gnu binary does not

@ShaneKilkelly I suppose the alternative would be to set default values in `start.sh`. It seems cleaner to do it in the config file though.

File: conf/envsubst_template.json (hunk 1 above)
Comment (JuneKelly) [reply conf/envsubst_template.json]: Ah, good catch 👍

File: Dockerfile [hunk 2]
```diff
@@ -26,9 +30,17 @@ RUN apt-get update && apt-get install -y git \
...
+
+COPY conf/envsubst_template.json envsubst_template.json
+COPY start.sh start.sh
+
+RUN mkdir conf
+RUN chown node:node conf
+
 USER node
+
+ENTRYPOINT ["/start.sh"]
```
Comment (gh2k) [inline Dockerfile]: One comment on this: We could leave `ENTRYPOINT` as-is and swap this out for `CMD`. I think `ENTRYPOINT` might lean towards the preferred way of doing things, but it's less consistent with how we run other apps in production.

In particular, we tend to run `kubectl exec <podname> bash` to get into the container and debug what's going on - which IIRC passes the command to the entrypoint.

**Review by gh2k — COMMENTED** [review]
LGTM - just one comment about `ENTRYPOINT`.

File: conf/envsubst_template.json (hunk 1 above)
Comment (mans0954) [reply conf/envsubst_template.json]: Done.

File: Dockerfile (hunk 2 above)
Comment (mans0954) [reply Dockerfile]: Done.

**Review by gh2k — APPROVED** [review]
Great. Thanks! :-)

**Pull request merged**